sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodegenExpressionCachingSuite.scala

```diff
@@ -85,6 +85,36 @@ class CodegenExpressionCachingSuite extends SparkFunSuite {
     assert(instance2.eval(null))
   }
 
+  test("SPARK-32903: GeneratePredicate should eliminate sub-expressions") {
+    Seq(true, false).foreach { enabled =>
+      val leaf1 = ExprWithEvaluatedState()
+      val leaf2 = ExprWithEvaluatedState()
+      val leaf3 = ExprWithEvaluatedState()
+      val leaf4 = ExprWithEvaluatedState()
+
+      val expr1 = And(leaf1, leaf2)
+      val expr2 = And(leaf3, leaf4)
+      val cond = Or(expr1, expr2)
```

Comment:
To avoid the misleading, shall we change like this?
```scala
- val expr1 = And(leaf1, leaf2)
- val expr2 = And(leaf3, leaf4)
- val cond = Or(expr1, expr2)
val cond = Or(And(leaf1, leaf2), And(leaf3, leaf4))
```

Reply:
`expr1` and `expr2` is not reused in the other places.

Reply:
Yeah, sure. I put it there just for easily explanation in comment.

Reply:
Changed. And also changed the comment.